src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -86,11 +86,172 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
+                        var nextChild = cSharpCodeBlock.Children[next];
+
+                        openBraceNode = previousChild;
+                        openBraceRange = previousChild.GetRangeWithoutWhitespace(source);
+                        closeBraceNode = nextChild;
+                        closeBraceRange = nextChild.GetRangeWithoutWhitespace(source);
+                    }
+                }
+            }
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            else if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
+                statement.Body is CSharpStatementBodySyntax csharpStatementBody)
+            {
+                openBraceNode = csharpStatementBody.OpenBrace;
+                openBraceRange = csharpStatementBody.OpenBrace.GetRange(source);
+                codeNode = csharpStatementBody.CSharpCode;
+                codeRange = csharpStatementBody.CSharpCode.GetRangeWithoutWhitespace(source);
+                closeBraceNode = csharpStatementBody.CloseBrace;
+                closeBraceRange = csharpStatementBody.CloseBrace.GetRange(source);
+            }
+            // @functions
+            // {
+            // }
+            else if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.First(c => c.Kind == SyntaxKind.RazorMetaCode);
+                var closeBrace = cSharpCode.Children.Last(c => c.Kind == SyntaxKind.RazorMetaCode);
+                var code = cSharpCode.Children.First(c => c.Kind == SyntaxKind.CSharpCodeBlock) as CSharpCodeBlockSyntax;
+
+                openBraceNode = openBrace;
+                openBraceRange = openBrace.GetRange(source);
+                codeNode = code!;
+                codeRange = code.GetRangeWithoutWhitespace(source);
+                closeBraceNode = closeBrace;
+                closeBraceRange = closeBrace.GetRange(source);
+            }
+            else
+            {
+                // This isn't a node we care about
+                return;
+            }
+
+            if (openBraceRange is not null &&
+                codeRange is not null &&
+                openBraceRange.End.Line == codeRange.Start.Line)
+            {
+                var indentation = context.Indentations[openBraceRange.Start.Line];
+                var desiredIndentationLevel = (indentation.HtmlIndentationLevel * (int)context.Options.TabSize) + (indentation.RazorIndentationLevel * (int)context.Options.TabSize);
```

Comment:
This looks odd. Multiplying level by tabsize means the result is not the "level" any more, but rather the number of characters. I think there is a method on `context` that gives you this, and handles the options correctly.

Reply:
I was able to simplify this code significantly, but none of the functions I found do the Html+Razor calculations for you.